### PR TITLE
SALTO-6592: fix elementsSource has implementation

### PR DESCRIPTION
**Pull request by shayc331**
_fix elementsSource has implementation_

---

_Additional context for reviewer_
* it is a fix for the change from this [pr](https://github.com/salto-io/salto/pull/6505). I changed `elementsSource.has` implementation to use `elementsSource.get` inside, which may cause performance issues.

---
_Release Notes_: 
_None_

---
_User Notifications_: 
_None_

### CODE REVIEW

File: packages/adapter-utils/src/element_source.ts [hunk 1]
```diff
@@ -67,10 +67,14 @@ export const buildElementsSourceFromElements = (
     }
   }
 
-  const get = async (id: ElemID): Promise<Value> => {
+  const getElement = (id: ElemID): Value => {
     const { parent } = id.createTopLevelParentID()
     const topLevelElement = elementsMap[parent.getFullName()]
-    const element = elementsMap[id.getFullName()] ?? (topLevelElement && resolvePath(topLevelElement, id))
+    return elementsMap[id.getFullName()] ?? (topLevelElement && resolvePath(topLevelElement, id))
```
Comment (ori-moisis) [inline packages/adapter-utils/src/element_source.ts:73]: `resolvePath` can handle the ID being the top level ID, no need to perform another lookup in the elements map
```suggestion
    return topLevelElement && resolvePath(topLevelElement, id)
```

File: packages/adapter-utils/src/element_source.ts [hunk 2]
```diff
@@ -67,10 +67,14 @@ export const buildElementsSourceFromElements = (
     }
   }
 
-  const get = async (id: ElemID): Promise<Value> => {
+  const getElement = (id: ElemID): Value => {
```
Comment (ori-moisis) [inline packages/adapter-utils/src/element_source.ts:70]: this doesn't return an element, the name is misleading

File: packages/adapter-utils/src/element_source.ts [hunk 3]
```diff
@@ -80,7 +84,8 @@ export const buildElementsSourceFromElements = (
       .find(values.isDefined)
   }
 
-  const has = async (id: ElemID): Promise<boolean> => (await get(id)) !== undefined
+  const has = async (id: ElemID): Promise<boolean> =>
+    getElement(id) !== undefined || awu(fallbackSources).some(async source => source.has(id))
```
Comment (ori-moisis) [inline packages/adapter-utils/src/element_source.ts:88]: nit: redundant `async` definition
```suggestion
    getElement(id) !== undefined || awu(fallbackSources).some(source => source.has(id))
```
this actually has an effect on performance in large enough numbers, not the case here, but still...

File: packages/adapter-utils/src/element_source.ts (hunk 1 above)
Comment (dantal4) [reply packages/adapter-utils/src/element_source.ts:73]: Theoretically, `elements` may contain fields without their parent types.

Comment (ori-moisis) [reply packages/adapter-utils/src/element_source.ts:73]: really?
if that is the case, then the code should have taken `baseID`

Comment (ori-moisis) [reply packages/adapter-utils/src/element_source.ts:73]: AFAICT all places in the code only pass in top level elements.

maybe worth adding a warn log if someone tries to build an element source here with something that isn't a top level element?

Comment (dantal4) [reply packages/adapter-utils/src/element_source.ts:73]: I prefer checking them both, as I see a situation when somebody uses this function in deploy, where changes have fields in them.
```typescript
    const baseId = id.createBaseID().parent
    const topLevelParentId = id.createTopLevelParentID().parent
    const element = elementsMap[baseId.getFullName()] ?? elementsMap[topLevelParentId.getFullName()]
    return element && resolvePath(element, id)
```

Comment (shayc331) [reply packages/adapter-utils/src/element_source.ts:73]: I agree with @dantal4

**Review by ori-moisis — COMMENTED** [review]
please fill in the PR description

File: packages/adapter-utils/src/element_source.ts [hunk 4]
```diff
@@ -67,10 +67,15 @@ export const buildElementsSourceFromElements = (
...
+    const topLevelParentId = id.createTopLevelParentID().parent
+    const element = elementsMap[baseId.getFullName()] ?? elementsMap[topLevelParentId.getFullName()]
+    return element && resolvePath(element, id)
+  }
+
   const get = async (id: ElemID): Promise<Value> => {
-    const { parent } = id.createTopLevelParentID()
-    const topLevelElement = elementsMap[parent.getFullName()]
-    const element = elementsMap[id.getFullName()] ?? (topLevelElement && resolvePath(topLevelElement, id))
+    const element = getValueByID(id)
```
Comment (ori-moisis) [inline packages/adapter-utils/src/element_source.ts:78]: nit: not an element

**Review by ori-moisis — APPROVED** [review]

**Pull request merged**